based/crates/reth/src/driver.rs

```diff
@@ -0,0 +1,345 @@
+        self.tx.send(Cmd::EnvV0 { env, resp: resp_tx }).await?;
+        resp_rx.await?.into_result()
+    }
+
+    /// Executes and records a fragment against the current unsealed block.
+    pub async fn new_frag_v0(&self, frag: FragV0) -> Result<FragStatus, DriverError> {
+        let (resp_tx, resp_rx) = oneshot::channel();
+        self.tx.send(Cmd::NewFragV0 { frag, resp: resp_tx }).await?;
+        resp_rx.await?.into_result()
+    }
+
+    /// Validates and finalizes the current unsealed block using the provided seal.
+    pub async fn seal_frag_v0(&self, seal_v0: SealV0) -> Result<(), DriverError> {
+        let (resp_tx, resp_rx) = oneshot::channel();
+        self.tx.send(Cmd::SealFragV0 { resp: resp_tx, seal: seal_v0 }).await?;
+        resp_rx.await?.into_result()
+    }
+
+    /// Notifies the driver about a forkchoice update and resets state on mismatch.
+    pub async fn forkchoice_updated(&self, new_block_number: u64) -> Result<(), DriverError> {
+        let (resp_tx, resp_rx) = oneshot::channel();
+        self.tx.send(Cmd::ForkchoiceUpdated { new_block_number, resp: resp_tx }).await?;
+        resp_rx.await?.into_result()
+    }
+
+    /// Returns a best-effort view of the current unsealed header used as "latest" when enabled.
+    pub async fn header_view(&self) -> HeaderView {
+        let (resp_tx, resp_rx) = oneshot::channel();
+
+        if self.tx.send(Cmd::GetHeaderView { resp: resp_tx }).await.is_err() {
+            return HeaderView { enabled: false, header: None };
+        }
+
+        match resp_rx.await {
+            Ok(reply) => reply.into_result().unwrap_or(HeaderView { enabled: false, header: None }),
+            Err(_) => HeaderView { enabled: false, header: None },
+        }
+    }
+}
+
+impl<E: UnsealedExecutor> DriverInner<E> {
+    fn reset_current_unsealed_block(&mut self) {
+        self.current_unsealed_block = None;
+        self.exec.reset();
+        self.fcu_count_since_unseal_reset = 0;
+    }
+
+    async fn handle_env_v0(&mut self, env: EnvV0) -> Result<(), DriverError> {
+        info!(block = env.number, "envV0 received");
+
+        if let Some(current) = self.current_unsealed_block.as_ref() {
+            let current_num = current.env.number;
+
+            if current_num >= env.number {
+                return Err(DriverError::UnsealedBlockInProgress { current: current_num, incoming: env.number });
+            }
+
+            info!(old = current_num, new = env.number, "env advanced, resetting");
+            self.reset_current_unsealed_block();
+        }
+
+        self.exec.ensure_env(&env).await?;
+        self.current_unsealed_block = Some(UnsealedBlock::new(env));
+        self.fcu_count_since_unseal_reset = 0;
+        Ok(())
+    }
+
+    async fn handle_forkchoice_updated(&mut self, new_block_number: u64) -> Result<(), DriverError> {
+        self.fcu_count_since_unseal_reset += 1;
+
+        let Some(ub) = self.current_unsealed_block.as_ref() else {
```

Comment:
Could you add references to the original based op geth code in a comment? So it's easier to cross-reference the implementations